packages/autorest.go/src/gocodemodel/gocodemodel.ts

```diff
@@ -0,0 +1,1252 @@
+export function isLiteralValueType(type: PossibleType): type is LiteralValueType {
+  return isConstantType(type) || isPrimitiveType(type);
+}
+
+export function isPrimitiveType(type: PossibleType): type is PrimitiveType {
+  return (<PrimitiveType>type).typeName !== undefined && !isStandardType(type) && !isTimeType(type);
+}
+
+export function isStandardType(type: PossibleType): type is StandardType {
+  return (<StandardType>type).packageName !== undefined && !isTimeType(type);
+}
+
+export function isTimeType(type: PossibleType): type is TimeType {
+  return (<TimeType>type).dateTimeFormat !== undefined;
+}
+
+export function isMapType(type: PossibleType): type is MapType {
+  return (<MapType>type).valueType !== undefined;
+}
+
+export function isModelType(type: PossibleType): type is ModelType {
+  return (<ModelType>type).format !== undefined;
+}
+
+export function isPolymorphicType(type: PossibleType): type is PolymorphicType {
+  return (<PolymorphicType>type).interface !== undefined;
+}
+
+export function isSliceType(type: PossibleType): type is SliceType {
+  return (<SliceType>type).elementType !== undefined;
+}
+
+export function isInterfaceType(type: PossibleType): type is InterfaceType {
+  return (<InterfaceType>type).possibleTypes !== undefined;
+}
+
+export function isLiteralValue(type: PossibleType): type is LiteralValue {
+  return (<LiteralValue>type).literal !== undefined;
+}
+
+export function getLiteralValueTypeName(literal: LiteralValueType): string {
+  if (isBytesType(literal)) {
+    return '[]byte';
+  } else if (isConstantType(literal)) {
+    return literal.name;
+  } else {
+    return literal.typeName;
+  }
+}
+
+export function getTypeDeclaration(type: PossibleType, pkgName?: string): string {
+  if (isPrimitiveType(type) || isStandardType(type)) {
+    return type.typeName;
+  } else if (isConstantType(type) || isInterfaceType(type) || isModelType(type) || isPolymorphicType(type)) {
+    if (pkgName) {
+      return `${pkgName}.${type.name}`;
+    }
+    return type.name;
+  } else if (isBytesType(type)) {
+    return '[]byte';
+  } else if (isLiteralValue(type)) {
+    return getTypeDeclaration(type.type, pkgName);
+  } else if (isMapType(type)) {
+    let pointer = '*';
+    if (type.valueTypeByValue) {
+      pointer = '';
+    }
+    return `map[string]${pointer}` + getTypeDeclaration(type.valueType, pkgName);
+  } else if (isSliceType(type)) {
+    let pointer = '*';
+    if (type.elementTypeByValue) {
+      pointer = '';
+    }
+    return `[]${pointer}` + getTypeDeclaration(type.elementType, pkgName);
+  } else if (isTimeType(type)) {
+    return 'time.Time';
+  } else {
+    throw new Error(`unhandled type ${typeof(type)}`);
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////////////////
+
+export class Info implements Info {
+  constructor(title: string) {
```

Comment:
JS has such nice object-literal initialization. Do you even need constructors?

Reply:
fixed

Reply:
This dovetails with the whole "null array vs empty array" conversation (that was my thinking anyways).  They might be able to be removed.  What's the idiom in TS when one should/shouldn't write a constructor for a type?